src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -0,0 +1,1282 @@
+                    byte status1 = value[0], status2 = value[1], status3 = value[2];
+                    if (!IsDigit(status1) || !IsDigit(status2) || !IsDigit(status3))
+                    {
+                        throw new HttpRequestException(SR.net_http_invalid_response);
+                    }
+
+                    _response.SetStatusCodeWithoutValidation((HttpStatusCode)(100 * (status1 - '0') + 10 * (status2 - '0') + (status3 - '0')));
+                }
+                else
+                {
+                    if (!HeaderDescriptor.TryGet(name, out HeaderDescriptor descriptor))
+                    {
+                        // Invalid header name
+                        throw new HttpRequestException(SR.net_http_invalid_response);
+                    }
+
+                    string headerValue = descriptor.GetHeaderValue(value);
+
+                    // Note we ignore the return value from TryAddWithoutValidation; 
+                    // if the header can't be added, we silently drop it.
+                    if (descriptor.HeaderType == HttpHeaderType.Content)
+                    {
+                        _response.Content.Headers.TryAddWithoutValidation(descriptor, headerValue);
+                    }
+                    else
+                    {
+                        _response.Headers.TryAddWithoutValidation(descriptor, headerValue);
+                    }
+                }
+            }
+
+            public void OnResponseHeadersComplete(bool endStream)
+            {
+                if (endStream)
+                {
+                    _responseComplete = true;
+                }
+
+                TaskCompletionSource<bool> readDataAvailable = _responseDataAvailable;
+                _responseDataAvailable = null;
+                readDataAvailable.SetResult(true);
+            }
+
+            public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
+            {
+                TaskCompletionSource<bool> readDataAvailable = null;
+
+                lock (_syncObject)
+                {
+                    if (_disposed)
+                    {
+                        return;
+                    }
+
+                    Debug.Assert(!_responseComplete);
+
+                    _responseBuffer.EnsureAvailableSpace(buffer.Length);
+                    buffer.CopyTo(_responseBuffer.AvailableSpan);
+                    _responseBuffer.Commit(buffer.Length);
+
+                    if (endStream)
+                    {
+                        _responseComplete = true;
+                    }
+
+                    if (_responseDataAvailable != null)
+                    {
+                        readDataAvailable = _responseDataAvailable;
+                        _responseDataAvailable = null;
+                    }
+                }
+
+                if (readDataAvailable != null)
+                {
+                    readDataAvailable.SetResult(true);
+                }
```

Comment:
I assume you're doing this completion outside of the lock in order to avoid invoking the continuations while the lock is held.  You can also achieve that by passing TaskCreationOptions.RunContinuationsAsynchronously to the TaskCompletionSource's ctor.